### DISCUSSION TITLE
How to separate my project into multiple parts

### DISCUSSION THREAD

**Discussion by FranciscoOssian** [premake/premake-core#2208] [Q&A] [answered]
Hello, can anyone help, please?
I have this Premake file:
```lua
workspace "MyProject"
    configurations { "Debug", "Release" }
    location "build"

project "MyProject"
    kind "ConsoleApp"
    language "C++"
    targetdir "build/bin/%{cfg.buildcfg}"
    objdir "build/obj/%{cfg.buildcfg}"

    files { "src/*.cpp", "include/*.hpp" }
    includedirs { "include", "vendor/eigen", "vendor/viennaCL", "vendor/viennaCL/CL" }

    defines { "VIENNACL_WITH_OPENCL" }

    filter "configurations:Debug"
        symbols "On"
        optimize "Off"

    filter "configurations:Release"
        symbols "Off"
        optimize "On"

    filter {}

    links { "OpenCL" }
```
And I want to create a `premake.lua` file in `vendor/eigen` and in `vendor/viennaCL`, but all attempts fail.

Even if I try to do it together and create it in the same file:

```lua
project "Eigen"
    kind "StaticLib"
    language "C++"
    includedirs { "vendor/eigen" }
```
and call it in the project "MyProject":

`links { "OpenCL", "Eigen" }`

The result is:

```bash
../src/main.cpp:1:10: fatal error: Eigen/Dense: No such file or directory
    1 | #include <Eigen/Dense>
      |          ^~~~~~~~~~~~~
```

Edit:

The project structure of this issue is here [https://github.com/FranciscoOssian/cpp-template](https://github.com/FranciscoOssian/cpp-template)

```bash
@:~/repos/cpp-template$ tree -L 2
.
├── include
│   └── test.hpp
├── premake.lua
├── src
│   ├── main.cpp
│   └── test.cpp
└── vendor
    ├── eigen
    └── viennaCL
```

The idea is simply to call the premake of the "X" dependency (creating file in "vendor/eigen") and configure it in the main premake, without writing details in the main premake.

**Comment by nickclark2016** [comment]
Have you tried using absolute paths? You can get the workspace location from the `wks` token. https://premake.github.io/docs/Tokens

**Reply by FranciscoOssian** [reply]
I tried using it for testing here, but I couldn't. For example, `wks` returns nil if I print it. And I'm not sure if I did it right because it gives the same error even if I don't use it. The biggest issue is that I imagine I shouldn't even need this because initially I'm trying to separate it in the same file, and yet it still gives the mentioned error.

How could I split it? Or how would I use this token if it did work? I really couldn't figure it out, sorry. Any help or suggestions would be greatly appreciated. Thank you in advance!

**Comment by samsinsane** [comment]
`includedirs` is for `#include "Header.h"`
`externalincludedirs` is for `#include <Header.h>`

You might find that `includedirs` works for `#include <>` and `externalincludedirs` works for `#include ""` for some systems/actions/compilers/whatever, but I would use the above as a rule of thumb when wanting to specify include directories. I thought this was covered in the "Write Premake Scripts" section of the documentation, but I can't see anything that covers it, we should definitely improve that. Similarly, [`includedirs`](https://premake.github.io/docs/includedirs) should reference [`externalincludedirs`](https://premake.github.io/docs/externalincludedirs/) so that users are able to find this related API.

**Reply by FranciscoOssian** [reply]
I just discovered that there's `-isystem` in compiler options; it might help because it suppresses some warnings related to third-party libraries. I tested using it in the context of my issue, but I'm still facing the same error.

**Reply by samsinsane** [reply]
`externalincludedirs` should map to `-isystem` for you.

So, to confirm, this is your directory structure:
```
premake5.lua
main.cpp
vendors/eigen/Eigen/Dense
```
and you've got this in your `premake5.lua`:
```lua
...
project "MyProject"
  ...
  externalincludedirs { "vendors/eigen" }
  ...
```
and in `main.cpp`:
```c++
#include <Eigen/Dense>
...
```

**Reply by FranciscoOssian** [reply]
The idea is to create some way of separating configurations into different files. Changing to `externalincludedirs` may alter the priority and warnings of this library, from what I researched. I updated my question to make it clearer.

**Comment by FranciscoOssian** [comment, marked as answer]
I found an example of premake that deals with GLFW and OpenGL, and the approach used is different from what I was trying, maybe that's why I always made mistakes. For now, the way I managed to solve it was simply by creating a function in the separate file with the configurations I wanted. For third parties static dependencies, this will be useful. It raised other questions for me, but I'll save them to research, and if I can't find answers, I'll ask another question. Thanks to those who were willing to help.

```lua
function useEigen()
    includedirs { "vendor/eigen" }
end
```

```lua
dofile "vendor/eigen/premake.lua"
project "Project"
    useEigen()
```
